File: js/test_money.js
const assert = require('assert')


class Money {
    constructor(amount, currency) {
        this.amount = amount
        this.currency = currency
    }

    times(multiplier) {
        return new Money(this.amount * multiplier, this.currency)
    }

    divide(divisor) {
        return new Money(this.amount / divisor, this.currency)
    }
}

const fiver = new Money(5, "USD")
const tenDollars = new Money(10, "USD")
assert.deepStrictEqual(fiver.times(2), tenDollars)

tenEuros = new Money(10, "EUR")
twentyEuros = tenEuros.times(2)
assert.deepStrictEqual(tenEuros.times(2), twentyEuros)
assert.strictEqual(twentyEuros.currency, "EUR")

originalMoney = new Money(4002, "KRW")
actualMoneyAfterDivision = originalMoney.divide(4)
expectedMoneyAfterDivision = new Money(1000.5, "KRW")
assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision)

File: py/test_money.py
import unittest


class Money:
    def __init__(self, amount, currency):
        self.amount = amount
        self.currency = currency

    def times(self, multiplier):
        return Money(self.amount * multiplier, self.currency)

    def divide(self, divisor):
        return Money(self.amount / divisor, self.currency)

    def __eq__(self, other):
        return self.amount == other.amount and self.currency == other.currency



class TestMoney(unittest.TestCase):
    def testMultiplicationInDollars(self):
        fiver = Money(5, "USD")
        ten = Money(10, "USD")
        tenner = fiver.times(2)
        self.assertEqual(ten, fiver.times(2))

    def testMultiplicationInEuros(self):
        tenEuros = Money(10, "EUR")
        twentyEuros = Money(20, "EUR")
        self.assertEqual(twentyEuros, tenEuros.times(2))
        self.assertEqual("EUR", twentyEuros.currency)

    def testDivision(self):
        originalMoney = Money(4002, "KRW")
        expectedMoneyAfterDivision = Money(1000.5, "KRW")
        self.assertEqual(expectedMoneyAfterDivision,
                         originalMoney.divide(4) )
        self.assertEqual(expectedMoneyAfterDivision.currency,
                         originalMoney.currency)


if __name__ == '__main__':
    unittest.main()
